### PR TITLE
New function recursive_fill in image_ops

### PR DESCRIPTION
**Pull request by jmlipman**
Implementation of a very common post-processing operation in image processing in a efficient manner embedded in a tensorflow graph. This function will iteratively fill the "holes" of a binary 2D or 3D image until it cannot fill more. It is also possible to set the sensitivity of the method by establishing a threhsold and a maximum number of iterations. For instance, let the threshold be 8 (or actually any number below or equal to 8)

Then from this:
0 0 0 0 0
0 1 1 1 0
0 1 0 1 0
0 1 1 1 0
0 0 0 0 0

We will obtain this:
0 0 0 0 0
0 1 1 1 0
0 1 1 1 0
0 1 1 1 0
0 0 0 0 0

### CODE REVIEW

**Comment by jmlipman** [comment]
I signed it!

File: tensorflow/contrib/image/python/ops/image_ops.py [hunk 1]
```diff
@@ -508,6 +511,82 @@ def has_zero():
...
+def recursive_fill(mask, threshold, total_dims, max_iter=None):
+  """Recursively iterates until the provided 0-1s 2D or 3D mask is filled.
+  If max_iter is provided, it will iterate for a number of times.
+  If total_dims is 2, threshold is recommended to be 4.
+  If total_dims is 3, threshold is recommended to be 13.
+
+  Args:
+    mask: A 2D or 3D `Tensor` of 0s and 1s representing the mask to be filled.
+    threshold: A threshold indicating sensitive the hole detector will be.
+    total_dims: Number of dimensions of the mask. Only 2 or 3 allowed.
```
Comment (ringw) [inline tensorflow/contrib/image/python/ops/image_ops.py]: We would normally get the static shape of the mask and require that the number of dimensions is set, and not pass the dimensionality as a separate argument. For example, see around line 250 of this file. Then if the number of dimensions is statically unknown, the caller can manually set it with e.g. `mask.set_shape([None, None])` for 2D.

File: tensorflow/contrib/image/python/ops/image_ops.py [hunk 2]
```diff
@@ -508,6 +511,82 @@ def has_zero():
...
       return components
 
+def recursive_fill(mask, threshold, total_dims, max_iter=None):
+  """Recursively iterates until the provided 0-1s 2D or 3D mask is filled.
+  If max_iter is provided, it will iterate for a number of times.
+  If total_dims is 2, threshold is recommended to be 4.
+  If total_dims is 3, threshold is recommended to be 13.
+
+  Args:
+    mask: A 2D or 3D `Tensor` of 0s and 1s representing the mask to be filled.
```
Comment (ringw) [inline tensorflow/contrib/image/python/ops/image_ops.py]: We should always allow 4-dimensional tensors, where the 0th axis contains a batch of images to be processed independently. The docstring should also explicitly state that a 3D or 4D input is always treated as containing voxels along the last 3 dimensions, which are always filled in 3-dimensional space.

**Review by ringw — CHANGES_REQUESTED** [review]
Thanks!

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @ringw: It has been 24 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by caisq** [comment]
Ping @jmlipman . Can you address the comments from @ringw?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 20 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ringw** [comment]
@jmlipman Hmm, I actually think conv2d and conv3d will already treat the 0th axis as a batch of independent images, so you shouldn't really need to modify the loop. You can treat a 4-D input as a batch of 3-D voxel masks if you just skip `array_ops.expand_dims(mask, 0)` in that case. Please also just copy `test_random_3d_fills` and create a 4-D test case!

File: tensorflow/contrib/image/python/ops/image_ops.py [hunk 3]
```diff
@@ -508,6 +511,82 @@ def has_zero():
...
+    center_index = math_ops.cast(filter_cube_size/2, dtypes.int32)
+    filt = state_ops.scatter_update(filt, center_index, -99)
+    filt = array_ops.reshape(filt, [3 for _ in range(total_dims)]+[1, 1])
+
+    mask = array_ops.expand_dims(mask, 0)
+    mask = array_ops.expand_dims(mask, -1)
+    new_voxels = array_ops.ones_like(mask)
+
+    # Note: the "y" argument needs to be given because as while_loop returns
+    # two elements, it requires two inputs as well.
```
Comment (ringw) [inline tensorflow/contrib/image/python/ops/image_ops.py]: What happens if you remove the second argument, and `loop_body` and `loop_cond` just reference `new_voxels` from the outer scope?

File: tensorflow/contrib/image/python/ops/image_ops.py [hunk 4]
```diff
@@ -508,6 +511,82 @@ def has_zero():
...
+    def loop_body(x, y): # pylint: disable=unused-argument
```
Comment (ringw) [inline tensorflow/contrib/image/python/ops/image_ops.py]: Maybe rename "x" to "current_mask"

**Review by ringw — CHANGES_REQUESTED** [review]

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @caisq: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by caisq** [comment]
pinging PR author @jmlipman

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tensorflowbutler** [comment]
It has been 43 days that this pull-request has stalled. Please create a new pull-request with the requested changes.

**Pull request closed without merging**